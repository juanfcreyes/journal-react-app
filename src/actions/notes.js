import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotesDB } from "../helpers/loadNotesDB";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        try {
            const doc = await db.collection(`${auth.uid}/journal/notes`).add(newNote);
            dispatch(setActiveNote(doc.id, newNote));
            dispatch(addNewNote({ id: doc.id, ...newNote}));
        } catch (error) {
            console.error(error);
        }
    
    }
};

export const startLoadNotes = () => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        try {
            const notes = await loadNotesDB(auth.uid);
            dispatch(loadNotes(notes));
        } catch (error) {
            dispatch(loadNotes([]));
            console.log(error);
        }
    }
};

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        const noteFS = {...note};
        delete noteFS.id;
        try {
            await db.doc(`${auth.uid}/journal/notes/${note.id}`).update(noteFS);     
        } catch (error) {
            console.log('startSaveNote error', error)
        }
        dispatch(refreshNote(note.id, noteFS));
        dispatch(setActiveNote(note.id, noteFS))
        Swal.fire('Save', note.title, 'success');
    }
}

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: note } = getState().notes;
        showUploadAlert();
        try {
            const fileUrl = await fileUpload(file);
            note.url = fileUrl;        
        } catch (error) {
            
        }
        dispatch(startSaveNote(note))
        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        try {
            const { auth } = getState();
            await db.doc(`${auth.uid}/journal/notes/${id}`).delete();
            Swal.fire('Deleted', 'Note deleted', 'info');
            dispatch(deleteNote(id));
        } catch (error) {
            console.error(error);
        }
      
    }
}

const showUploadAlert = () => {
    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait',
        allowOutsideClick: false,
        willOpen: () => {
            Swal.showLoading()
        },
        showCloseButton: false,
        showCancelButton: false,
        showConfirmButton: false
    });
} 

export const addNewNote = (note) => ({
    type: types.notesAddnew,
    payload: {
        ...note
    }
});

export const setActiveNote = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
});

export const loadNotes = (notes) => ({
    type : types.notesLoad,
    payload: notes
});

export const refreshNote = (id, note) => ({
    type: types.notesUploaded,
    payload: {
        id,
        note: { id, ...note }
    }
});

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: {
        id
    }
});

export const notesLogout = () => ({
    type: types.notesCleaning,
});

