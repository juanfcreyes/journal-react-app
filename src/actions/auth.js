import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";
import { notesLogout } from "./notes";

export const startLogin = (email, password) => {
    return async (dispatch) => {
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password).then(({user}) => {
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
            dispatch(finishLoading());
            console.log('error', error);
        })
        
    }
}

export const startRegister = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(async ({user}) => {
            await user.updateProfile({displayName: name})
            dispatch(login(user.uid, user.displayName));
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
        });
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        });
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(notesLogout()); 
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});