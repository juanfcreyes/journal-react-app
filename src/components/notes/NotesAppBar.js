import React from 'react'
import moment from 'moment'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const {active: note} = useSelector(state => state.notes);
    const dispatch = useDispatch();
    const momentDate = moment(note.date);
    
    const handleSave = () => {
        dispatch(startSaveNote(note));
    }

    const handlePictureUpload = () => {
       document.getElementById('file_input').click();
    }

    const handleFileChange = (e) => {
       const file = e.target.files[0];
       if (file) {
           dispatch(startUploading(file));
       }
    }

    return (
        <div className='notes__appbar'>
            <span>{momentDate.format('dddd DD MMMM YYYY')}</span>
            <input type='file' style={ {display: 'none' }} onChange={handleFileChange} id='file_input' name='file_input'></input>
            <div>
                <button className='btn' onClick={handlePictureUpload}>picture</button>
                <button className='btn' onClick={handleSave} >Save</button>
            </div>
        </div>
    )
}
