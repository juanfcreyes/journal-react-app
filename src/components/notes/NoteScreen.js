import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { setActiveNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();
    const {active: note} = useSelector(state => state.notes);
    const [formValues, handleInputChange, reset] = useForm(note);
    const {title, body} = formValues;
    const activeId = useRef(note.id);

    useEffect(() => {
        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset, formValues]);

    useEffect(() => {
        dispatch(setActiveNote(formValues.id, {...formValues}));
    }, [formValues, dispatch]);

    const handleDelete = () => {
        dispatch(startDeleting(note.id));
    };
    
    return (
        <div className="notes__main-content animate__animated animate__fadeIn">
            <NotesAppBar></NotesAppBar>
            <div className='notes__content'>
                <input type='text' placeholder='Some awesome title' className='notes__title-input' autoComplete='off' name='title' onChange={handleInputChange} value={title}></input>
                <textarea placeholder='What happend today' className='notes__textarea' name='body' onChange={handleInputChange} value={body} > </textarea>
                { 
                    note.url && 
                    <div className='notes__image'>
                        <img src={note.url} alt='imagen de ejemplo' style={{height:'100%', width:'100%'}}></img>
                    </div>
                }
            </div>
            <button className='btn btn-danger' onClick={handleDelete}>Delete </button>
        </div>
    )
}
