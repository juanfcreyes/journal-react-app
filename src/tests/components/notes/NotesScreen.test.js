import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { setActiveNote } from "../../../actions/notes";
import { NoteScreen } from "../../../components/notes/NoteScreen";

jest.mock('../../../actions/notes', () => ({
    setActiveNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {

    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 'safdsgfd4g6fg',
            title: 'Titulo',
            body: 'cuerpo',
            date: 12348875413
        }
    }
}

let store = mockStore(initState);
store.dispatch = jest.fn();
describe('Pruebas sobre NotesScreen', () => {

    const wrapper = mount(
        <Provider store={store}>
            <NoteScreen></NoteScreen>
        </Provider>
    )
    

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de disparar el activeNode', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Nuevo titulo'
            }
        });

        expect(setActiveNote).toHaveBeenCalledWith(
            'safdsgfd4g6fg',
            { 
                id: 'safdsgfd4g6fg',
                title: 'Nuevo titulo',
                body: 'cuerpo',
                date: expect.any(Number)
            }
        );
    });
    
    
});
