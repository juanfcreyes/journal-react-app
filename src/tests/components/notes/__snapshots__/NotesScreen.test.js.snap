// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre NotesScreen Debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn"
>
  <div
    className="notes__appbar"
  >
    <span>
      Saturday 23 May 1970
    </span>
    <input
      id="file_input"
      name="file_input"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Titulo"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happend today"
      value="cuerpo"
    >
       
    </textarea>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete 
  </button>
</div>
`;
